python/lsst/rubintv/production/slac/botTesting.py

```diff
@@ -704,75 +725,89 @@ def run(self):
 
         Returns
         -------
-        records : `list` [`lsst.daf.butler.DimensionRecord`]
-            The list of records.
+        shards : `dict` [`lsst.daf.butler.DimensionRecord`, `list` [`str`]]
+            A dictionary, keyed by exposure record, of the binned image files.
         """
         # must grab this before scanning for data, to protect against new
         # images arriving during the scrape. This watcher watched for
         # binnedImages, which are always an earlier or equal expRecord to the
         # raw one, and so if the right one to choose
         mostRecentExp = self.watcher.getMostRecentExpRecord()
 
-        records = []
+        records = {}
         dataPath = self.locationConfig.calculatedDataPath
         expIds = getBinnedImageExpIds(dataPath, self.instrument)
+        if not expIds:
+            return {}
+
+        allFiles = getBinnedImageFiles(dataPath, self.instrument)
+        self.log.debug(f'Found {len(expIds)} expIds with binned images, and {len(allFiles)} associated files')
         for expId in expIds:
             record = getExpRecord(self.butler, self.instrument, expId=expId)
-            if findIncomplete:  # we don't need to check for completeness so skip the costly checks
-                records.append(record)
-                continue
-
-            expected = getNumExpectedItems(record)
-            binnedImages = getBinnedImageFiles(dataPath, self.instrument, expId=expId)
-            if expected == len(binnedImages):
-                self.log.debug(f"{expId} is complete with {len(binnedImages)} images found")
-                records.append(record)
-            else:
-                self.log.debug(f"{expId} is not complete, with {len(binnedImages)} items of {expected} found")
-                if findIncomplete:
-                    records.append(record)
+            # a list comp is *much* faster than re-globbing if there are a lot
+            # of files.
+            binnedImages = [f for f in allFiles if f.find(f'{expId}') != -1]
+            records[record] = binnedImages
 
         # Check to see if there is anything more recent than the most recent
         # expRecord which existed when we started looking for data, and if so,
         # remove them, so that we don't collide with the running processes.
-        records = [r for r in records if r.timespan.end < mostRecentExp.timespan.end]
+        records = {r: files for r, files in records.items() if r.timespan.end < mostRecentExp.timespan.end}
         return records
 
     def run(self):
         """Run continuously, looking for complete file sets and plotting them.
         """
-        # TODO: turn this into an end-of-day service, checking for day rollover
-        # and calling findLeftoverMosaics() with findIncomplete=True, and set
-        # deleteRegardless when plotting them to tidy up at the end of the day.
+        # workloads is a list of tuples, each of which is: (a filter function,
```

Comment:
This is where I say this should definitely be a (data)class.

Reply:
Yup, definitely agree. Could you just take a quick look at `c7882a5` and check that's what you meant? I think that making the items `Callable` and making the call to `partial()` in making the dataclasses is the right way to go, so that the calling code is basically unchanged, as opposed to putting the args and kwargs in the dataclass, but just wanted to check this is what you were thinking.